fs/vfs/fs_eventfd.c

```diff
@@ -0,0 +1,639 @@
+
+  /* TODO reimplement this with minor bit map ? */
+
+  return (minor++) & 511;
+}
+
+void release_minor(int minor)
+{
+}
+
+int eventfd_open(FAR struct file *filep)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct eventfd_priv_s *priv = inode->i_private;
+  int ret;
+
+  /* Get exclusive access to the device structures */
+
+  ret = nxsem_wait(&priv->exclsem);
+  if (ret < 0)
+    {
+      return -1;
+    }
+
+  finfo("crefs: %d <%s>\n", priv->crefs, inode->i_name);
+
+  if (priv->crefs >= 255)
+    {
+      /* More than 255 opens; uint8_t would overflow to zero */
+
+      ret = -EMFILE;
+    }
+  else
+    {
+      /* Save the new open count on success */
+
+      priv->crefs += 1;
+      ret = OK;
+    }
+
+  nxsem_post(&priv->exclsem);
+  return ret;
+}
+
+int eventfd_close(FAR struct file *filep)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct eventfd_priv_s *priv = inode->i_private;
+  FAR char *devpath;
+  int ret;
+
+  /* Get exclusive access to the device structures */
+
+  ret = nxsem_wait(&priv->exclsem);
+  if (ret < 0)
+    {
+      return -errno;
+    }
+
+  finfo("crefs: %d <%s>\n", priv->crefs, inode->i_name);
+
+  /* Decrement the references to the driver.  If the reference count will
+   * decrement to 0, then uninitialize the driver.
+   */
+
+  if (priv->crefs > 1)
+    {
+      /* Just decrement the reference count and release the semaphore */
+
+      priv->crefs -= 1;
+      nxsem_post(&priv->exclsem);
+      return OK;
+    }
+
+  /* Re-create the path to the driver. */
+
+  finfo("destroy\n");
+  ret = asprintf(&devpath, EVENTFD_FMT, priv->minor);
+  if (ret < 0)
+    {
+      _err("ERROR: Failed to allocate the driver path\n");
+      nxsem_post(&priv->exclsem);
+      return -ENOMEM;
+    }
+
+  /* Will be unregistered later after close is done */
+
+  unregister_driver(devpath);
+
+  free(devpath);
+
+  DEBUGASSERT(priv->exclsem.semcount == 0);
+  release_minor(priv->minor);
+  eventfd_destroy(priv);
+
+  return OK;
+}
+
+ssize_t eventfd_do_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct eventfd_priv_s *dev = inode->i_private;
+  ssize_t ret;
+  int sval;
+
+  if (len < sizeof(eventfd_t) || buffer == NULL)
+    {
+      return -EINVAL;
+    }
+
+  ret = nxsem_wait(&dev->exclsem);
+  if (ret < 0)
+    {
+      return -errno;
+    }
+
+  /* Wait for an incoming event */
+
+  while (dev->counter == 0)
+    {
+      if (filep->f_oflags & O_NONBLOCK)
+        {
+          nxsem_post(&dev->exclsem);
+          return -EAGAIN;
+        }
+
+      /* Wait for a writer to notify */
+
+      sched_lock();
+      nxsem_post(&dev->exclsem);
+      ret = nxsem_wait(&dev->rdsem);
+      sched_unlock();
+
+      if (ret < 0 || (ret = nxsem_wait(&dev->exclsem)) < 0)
+        {
+          return ret;
+        }
+    }
+
+  /* Device ready for read */
+
+  if (dev->mode_semaphore)
+    {
+      *(eventfd_t *)buffer = 1;
```

Comment:
All data pointer types in common code should be prefixed with the qualifier FAR (which is defined to be nothing in most cases).  This applies in several other places as well.